File: fries-leetcode-training/06-binary-tree/03-binary-tree-level-order-traversal.py
"""
LeetCode 102. 二叉树的层序遍历
https://leetcode.cn/problems/binary-tree-level-order-traversal/

给你二叉树的根节点root，返回其节点值的层序遍历。

BFS队列

时间复杂度：O(n)
空间复杂度：O(w) w为树的最大宽度
"""

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def level_order(root):
    """
    二叉树的层序遍历 - BFS队列法
    
    Args:
        root: 二叉树根节点
        
    Returns:
        层序遍历结果
    """
    if not root:
        return []
    
    result = []
    queue = [root]
    
    while queue:
        level_size = len(queue)
        level = []
        
        for _ in range(level_size):
            node = queue.pop(0)
            level.append(node.val)
            
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
        
        result.append(level)
    
    return result


def test_level_order():
    """测试函数"""
    # 创建测试二叉树: [3,9,20,null,null,15,7]
    #     3
    #    / \
    #   9  20
    #     /  \
    #    15   7
    
    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)
    
    result = level_order(root)
    print(f"层序遍历结果: {result}")
    # 期望: [[3], [9, 20], [15, 7]]


if __name__ == "__main__":
    test_level_order()

File: fries-leetcode-training/06-binary-tree/03-binary-tree-level-order-traversal.cpp
/**
 * LeetCode 102. 二叉树的层序遍历
 * https://leetcode.cn/problems/binary-tree-level-order-traversal/
 * 
 * 给你二叉树的根节点root，返回其节点值的层序遍历。
 * 
 * BFS队列
 * 
 * 时间复杂度：O(n)
 * 空间复杂度：O(w) w为树的最大宽度
 */

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> result;
        if (!root) {
            return result;
        }
        
        queue<TreeNode*> q;
        q.push(root);
        
        while (!q.empty()) {
            int level_size = q.size();
            vector<int> level;
            
            for (int i = 0; i < level_size; i++) {
                TreeNode* node = q.front();
                q.pop();
                level.push_back(node->val);
                
                if (node->left) {
                    q.push(node->left);
                }
                if (node->right) {
                    q.push(node->right);
                }
            }
            
            result.push_back(level);
        }
        
        return result;
    }
};

// 测试函数
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

void testLevelOrder() {
    Solution solution;
    
    // 创建测试二叉树: [3,9,20,null,null,15,7]
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);
    
    vector<vector<int>> result = solution.levelOrder(root);
    
    cout << "层序遍历结果: [";
    for (int i = 0; i < result.size(); i++) {
        cout << "[";
        for (int j = 0; j < result[i].size(); j++) {
            cout << result[i][j];
            if (j < result[i].size() - 1) cout << ",";
        }
        cout << "]";
        if (i < result.size() - 1) cout << ",";
    }
    cout << "]" << endl;  // 期望: [[3],[9,20],[15,7]]
}

int main() {
    testLevelOrder();
    return 0;
}
